src/common/grpc/src/channel_manager.rs

```diff
@@ -172,65 +206,67 @@ impl ChannelManager {
             endpoint = endpoint.connect_timeout(dur);
         }
-        if let Some(limit) = self.config.concurrency_limit {
+        if let Some(limit) = self.config().concurrency_limit {
             endpoint = endpoint.concurrency_limit(limit);
         }
-        if let Some((limit, dur)) = self.config.rate_limit {
+        if let Some((limit, dur)) = self.config().rate_limit {
             endpoint = endpoint.rate_limit(limit, dur);
         }
-        if let Some(size) = self.config.initial_stream_window_size {
+        if let Some(size) = self.config().initial_stream_window_size {
             endpoint = endpoint.initial_stream_window_size(size);
         }
-        if let Some(size) = self.config.initial_connection_window_size {
+        if let Some(size) = self.config().initial_connection_window_size {
             endpoint = endpoint.initial_connection_window_size(size);
         }
-        if let Some(dur) = self.config.http2_keep_alive_interval {
+        if let Some(dur) = self.config().http2_keep_alive_interval {
             endpoint = endpoint.http2_keep_alive_interval(dur);
         }
-        if let Some(dur) = self.config.http2_keep_alive_timeout {
+        if let Some(dur) = self.config().http2_keep_alive_timeout {
             endpoint = endpoint.keep_alive_timeout(dur);
         }
-        if let Some(enabled) = self.config.http2_keep_alive_while_idle {
+        if let Some(enabled) = self.config().http2_keep_alive_while_idle {
             endpoint = endpoint.keep_alive_while_idle(enabled);
         }
-        if let Some(enabled) = self.config.http2_adaptive_window {
+        if let Some(enabled) = self.config().http2_adaptive_window {
             endpoint = endpoint.http2_adaptive_window(enabled);
         }
-        if let Some(tls_config) = &self.client_tls_config {
+        if let Some(tls_config) = &self.inner.client_tls_config {
             endpoint = endpoint
                 .tls_config(tls_config.clone())
                 .context(CreateChannelSnafu)?;
         }
 
         endpoint = endpoint
-            .tcp_keepalive(self.config.tcp_keepalive)
-            .tcp_nodelay(self.config.tcp_nodelay);
+            .tcp_keepalive(self.config().tcp_keepalive)
+            .tcp_nodelay(self.config().tcp_nodelay);
 
         Ok(endpoint)
     }
 
     fn trigger_channel_recycling(&self) {
         if self
+            .inner
             .channel_recycle_started
             .compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed)
             .is_err()
         {
             return;
         }
 
-        let pool = self.pool.clone();
+        let pool = self.pool().clone();
+        let cancel = self.inner.cancel.clone();
         let _handle = common_runtime::spawn_global(async {
-            recycle_channel_in_loop(pool, RECYCLE_CHANNEL_INTERVAL_SECS).await;
+            recycle_channel_in_loop(pool, cancel, RECYCLE_CHANNEL_INTERVAL_SECS).await;
```

Comment:
how about just pass `Inner` into it and remove arc on both pool&channel_recycle_started?

Reply:
The Arc of channel_recycle_started can be removed. Pool not, since Pool will share between `Inner` and `channel recyle task`.

Reply:
> The Arc of channel_recycle_started can be removed. Pool not, since Pool will share between `Inner` and `channel recyle task`.

How about directly pass Arc<Inner> to recycle task? Since now most field of `Inner` is passed to `recycle_channel_in_loop` anyway

Reply:
This will cause the loop always hold a reference of inner, and then the drop logic of inner cannot be executed.

Reply:
> This will cause the loop always hold a Arc, and then the drop logic of inner cannot be executed.

how about a weak ref

Reply:
In general, I tend not to use weak refs.

Reply:
> In general, I tend not to use weak ref.

not using weak still sound semantically misfitting, a task that refresh stuffs for the pool shouldn't hold a strong count in isolation, this kind of break the tree model of data that Arc should be using(creating a root for tree that floating in the void), this is exactly where weak should be use I think, see more at https://effective-rust.com/references.html?highlight=Weak#smart-pointer-types